central/group/datastore/datastore_impl.go

```diff
@@ -70,18 +101,26 @@ func (ds *dataStoreImpl) Add(ctx context.Context, group *storage.Group) error {
 		return errox.InvalidArgs.CausedBy(err)
 	}
 
-	if group.GetProps().GetId() != "" {
-		return errox.InvalidArgs.Newf("id should be empty but %q was provided", group.GetProps().GetId())
+	if err := setGroupIDForNewGroup(group); err != nil {
+		return err
 	}
 
-	if group.GetProps() != nil {
-		group.GetProps().Id = GenerateGroupID()
-	} else {
-		// Theoretically should never happen, as the auth provider ID is required to be set.
-		group.Props = &storage.GroupProperties{Id: GenerateGroupID()}
+	// Lock to ensure synchronization between the get and upsert
+	ds.lock.Lock()
```

Comment:
Remove this lock on add.

Reply:
Can't. It fetches default groups (see line 112) to check before it adds. This was in a transaction before and since I can't use a transaction in the datastore nor use a check-and-set, I figured this is the best approach for the moment.